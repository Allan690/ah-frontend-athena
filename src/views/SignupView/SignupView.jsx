import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux'
import SignupForm from '../../components/Signup/SignupForm'
import { signUp } from '../../actions/signupAction/signupAction'

export class SignupView extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            username: '',
            email: '',
            password: '',
            errors: {}
        };

    }
    componentWillReceiveProps(nextProps){
        const errors = nextProps.data
        if(errors){
            this.setState({errors:errors});
        }

    }
    onChange = (e)=>{

        this.setState({[e.target.name]: e.target.value});
    }
    onSubmit = (e)=> {
        e.preventDefault();
        const userobj = this.state
        const user = {
            username: userobj.username,
            email: userobj.email,
            password: userobj.password
        }
        const { signUp } = this.props;
        signUp(user)
    }
    render() {
        const {errors} = this.state
        return (
          <div>
            <SignupForm
              onChange={this.onChange}
              onSubmit={this.onSubmit}
              onClick={this.onClick}
              errors={errors}
            />
          </div>

        );
    }
}

SignupView.propTypes = {
    data: PropTypes.shape({}),
    signUp: PropTypes.func

};

SignupView.defaultProps = {
    data: {},
    signUp: () => {}
  }
export const mapStateToProps = (state)=>{
    return {
        data: state.user.user
    }
}

export default connect(
    mapStateToProps,
    {signUp}
) (SignupView);