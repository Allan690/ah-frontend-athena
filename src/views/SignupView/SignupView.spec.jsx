import React from 'react'
import {shallow} from 'enzyme';
import {SignupView, mapStateToProps} from './SignupView';

const mockSignUpFn = jest.fn();
const initialState = {
    user:{user:{
        username: 'kasule',
        email: '',
        password: '',
        errors: {}
    }}
}
let wrapper;

describe("test signup view", () => {
    beforeEach(() => {
        wrapper = shallow(<SignupView {...mockSignUpFn} />)
    })
    it('should SignupComponent match the snapshot', ()=>{
        expect(wrapper).toMatchSnapshot();
    })
    it('test handle onsubmit event', () =>{
        wrapper.instance().onSubmit({preventDefault() {}})
        expect(wrapper.state()).toEqual({"email": "", "errors": {}, "password": "", "username": ""})
    })

    it('should handle onchange function', ()=>{
        wrapper.instance().onChange({target:{name:'email',value:'kasule@andela.com'}})
        expect(wrapper.state('email')).toEqual('kasule@andela.com')
    })

    it('should will recieve signup props', () => {
        wrapper.setProps({data: {errors: 'username should be between 3 to 8 characters'}})
        expect(wrapper.state('errors')).toEqual({'errors': 'username should be between 3 to 8 characters'})
    })

    it('should map state to props', () =>{
        expect(mapStateToProps(initialState).data['username']).toEqual('kasule')
    })
})
