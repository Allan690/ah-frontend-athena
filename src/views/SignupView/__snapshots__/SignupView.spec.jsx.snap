// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test signup view should SignupComponent match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignupView
    _isMockFunction={true}
    data={Object {}}
    getMockImplementation={[Function]}
    getMockName={[Function]}
    mock={
      Object {
        "calls": Array [],
        "instances": Array [],
        "invocationCallOrder": Array [],
        "results": Array [],
      }
    }
    mockClear={[Function]}
    mockImplementation={[Function]}
    mockImplementationOnce={[Function]}
    mockName={[Function]}
    mockRejectedValue={[Function]}
    mockRejectedValueOnce={[Function]}
    mockReset={[Function]}
    mockResolvedValue={[Function]}
    mockResolvedValueOnce={[Function]}
    mockRestore={[Function]}
    mockReturnThis={[Function]}
    mockReturnValue={[Function]}
    mockReturnValueOnce={[Function]}
    signUp={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <SignupForm
        errors={Object {}}
        onChange={[Function]}
        onSubmit={[Function]}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errors": Object {},
        "onChange": [Function],
        "onClick": undefined,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <SignupForm
          errors={Object {}}
          onChange={[Function]}
          onSubmit={[Function]}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errors": Object {},
          "onChange": [Function],
          "onClick": undefined,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
